### PR TITLE
[On hold] Autofetch

### PR DESCRIPTION
**Pull request by Drahcirius**
# Autofetch

Pull Request for #1376 

Autofetch is a new client setting that forces the lib to fetch  data if they were not originally cached. Usage is as follows: 
```js
const client = new Client({ autofetch: [
    'MESSAGE_CREATE',
    'MESSAGE_UPDATE',
    'MESSAGE_REACTION_ADD',
    'MESSAGE_REACTION_REMOVE',
] });
```
## Behaviour
- Check member cache for author and mentioned users. If members are not in cache, fetch guild members before emitting message event.
- Cache message and emit messageUpdate event instead of discarding update. `<message>.edits` will be empty because the old message was not cached.
- If the message that was reacted to was not previously in the cache, fetch the message and emit messageReactionAdd event instead of discarding the event.

If there are other events or properties that can be prefetched, let me know and I will include it. 

**Semantic versioning classification:**  
- [x] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: src/client/actions/MessageReactionAdd.js [hunk 1]
```diff
@@ -9,16 +9,30 @@ const Constants = require('../../util/Constants');
...
     const channel = this.client.channels.get(data.channel_id);
     if (!channel || channel.type === 'voice') return false;
     // Verify message
     const message = channel.messages.get(data.message_id);
-    if (!message) return false;
     if (!data.emoji) return false;
+    if (!message) {
+      if (this.client.options.autofetch.includes(Constants.WSEvents.MESSAGE_REACTION_ADD)) {
+        // Check if message is fetchable
+        if (!channel.permissionsFor(channel.guild.me).has('READ_MESSAGES')) return false;
```
Comment (SpaceEEC) [inline src/client/actions/MessageReactionAdd.js]: `READ_MESSAGE_HISTORY` is also required to fetch a message iirc.

File: src/client/actions/MessageReactionAdd.js [hunk 2]
```diff
@@ -9,16 +9,30 @@ const Constants = require('../../util/Constants');
...
+
+        try {
+          const fetchedMessage = await channel.fetchMessage(data.message_id);
+          const reaction = fetchedMessage._addReaction(data.emoji, user);
+          this.client.emit(Constants.Events.MESSAGE_REACTION_ADD, reaction, user);
```
Comment (SpaceEEC) [inline src/client/actions/MessageReactionAdd.js]: Shouldn't you return the message, the reaction and the user here like line 40 does it too?

File: src/client/actions/MessageReactionRemove.js [hunk 3]
```diff
@@ -9,16 +9,30 @@ const Constants = require('../../util/Constants');
...
     const channel = this.client.channels.get(data.channel_id);
     if (!channel || channel.type === 'voice') return false;
     // Verify message
     const message = channel.messages.get(data.message_id);
-    if (!message) return false;
     if (!data.emoji) return false;
+    if (!message) {
+      if (this.client.options.autofetch.includes(Constants.WSEvents.MESSAGE_REACTION_REMOVE)) {
+        // Check if message is fetchable
+        if (!channel.permissionsFor(channel.guild.me).has('READ_MESSAGES')) return false;
```
Comment (SpaceEEC) [inline src/client/actions/MessageReactionRemove.js]: Same as above (`READ_MESSAGE_HISTORY`)

File: src/client/actions/MessageReactionRemove.js [hunk 4]
```diff
@@ -9,16 +9,30 @@ const Constants = require('../../util/Constants');
...
+
+        try {
+          const fetchedMessage = await channel.fetchMessage(data.message_id);
+          const reaction = fetchedMessage._addReaction(data.emoji, user);
+          this.client.emit(Constants.Events.MESSAGE_REACTION_REMOVE, reaction, user);
```
Comment (SpaceEEC) [inline src/client/actions/MessageReactionRemove.js]: Shouldn't you also return here?

File: src/client/actions/MessageUpdate.js [hunk 5]
```diff
@@ -17,10 +18,23 @@ class MessageUpdateAction extends Action {
...
-      };
+      if (this.client.options.autofetch.includes(Constants.WSEvents.MESSAGE_UPDATE)) {
+        if (data.author) {
+          const newMessage = channel._cacheMessage(new Message(channel, data, client));
+          client.emit(Constants.Events.MESSAGE_UPDATE, null, newMessage);
+          return {
+            old: null,
+            updated: newMessage,
+          };
+        } else {
```
Comment (SpaceEEC) [inline src/client/actions/MessageUpdate.js]: Isn't that `else` redundant as you return above?

File: src/client/actions/MessageUpdate.js [hunk 6]
```diff
@@ -17,10 +18,23 @@ class MessageUpdateAction extends Action {
...
+          try {
+            const fetchedMessage = await channel.fetchMessage(data.id);
+            client.emit(Constants.Events.MESSAGE_UPDATE, null, fetchedMessage);
```
Comment (SpaceEEC) [inline src/client/actions/MessageUpdate.js]: Shouldn't you return here too?

File: jsdoc.json [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (devsnek) [inline jsdoc.json]: wtf is this?

**Comment by KingDGrizzle** [comment]
Don't want to necropost or whatever, but would it be fine if I "ported" or adapted this to the latest master, with the requested changes as well?

**Comment by iCrawl** [comment]
No. Because we still do not have a verdict on how it's going to be fully implemented.

**Comment by codeHusky** [comment]
Any updates? Would be nice if this, or something similar to this, was added to master.

**Comment by codeHusky** [comment]
Is this necessary after closing #3070

**Comment by appellation** [comment]
Autofetch is a complimentary feature to partials, although this PR was implemented without the consideration of partials and would need to be rewritten.

**Comment by Drahcirius** [comment]
After some discussion we've decided that #3070 will be the solution going forward 🎉

**Pull request closed without merging**